bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -83,16 +206,56 @@ struct SourceProtocol<T: AsyncRead + AsyncWrite + Unpin + Send> {
 
     /// Transport to the destination Instance.
     conn: WebSocketStream<T>,
+
+    /// Guest page table dirty bits to restore in the event of a migration
+    /// failure, so that a subsequent migration can attempt to offer only dirty
+    /// pages. These dirty bits are accumulated across all RAM push phases, so
+    /// that any pages which become dirty after the pre-pause RAM push are also
+    /// added to these bitmaps.
+    ///
+    /// This is `Some` if we should attempt to re-set dirty bits on guest pages
+    /// in the event of a migration failure, and `None` if we cannot do so. It
+    /// will be `Some` if (and only if):
+    ///
+    /// - The current bhyve version supports the `VM_NPT_OPERATION` ioctl (i.e.
+    ///   it is at least bhyve v17 or later),
+    /// - A previous attempt to re-dirty pages has failed (viz.
+    ///   [`PersistentState::has_retrying_ever_failed`]), in which case, we can
+    ///   no longer trust  that all previously dirtied pages were re-dirtied
+    ///   correctly.
+    ///
+    /// Otherwise, we must fall back to always offering all pages in the initial
+    /// pre-pause RAM push phase.
+    dirt: Option<std::collections::HashMap<GuestAddr, PageBitmap>>,
 }
 
+type PageBitmap = [u8; PAGE_BITMAP_SIZE];
```

Comment:
Nit: I know it's totally legal, but it surprises me a bit to see the constant defined on the line _after_ it's used. :-D

Reply:
happy to move it if you'd like

Reply:
I leave it to your capable discretion. It's just always a bit surprising to me when I encounter it.